testsuite/tests/quotation/quotation_typing.ml

```diff
@@ -51,141 +56,186 @@ type 'a t1 = 'a expr
 
 type 'a t2 = <['a]> expr;;
 [%%expect {|
-Line 52, characters 15-17:
-52 | type 'a t2 = <['a]> expr;;
+Line 57, characters 15-17:
+57 | type 'a t2 = <['a]> expr;;
                     ^^
-Error: Type variable "a" is used at Line 52, characters 15-17
-       in a context with one layer of quotation (<[ ... ]>);
-       it should only be used in a context with no quotations or splices.
+Error: Type variable "'a" is used at Line 57, characters 15-17,
+       within a quotation (<[ ... ]>); it already occurs outside quotations.
+       Hint: Consider using "$'a".
 |}];;
 
 type 'a t3 = $'a -> $'a -> 'a expr;;
 [%%expect {|
 Line 1, characters 27-29:
 1 | type 'a t3 = $'a -> $'a -> 'a expr;;
                                ^^
-Error: Type variable "'a" is used at Line 1, characters 27-29
-       in a context with no quotations or splices;
-       it should only be used in a context with one layer of splicing ($).
+Error: Type variable "'a" is used at Line 1, characters 27-29,
+       outside quotations;
+       it already occurs Uncaught exception: Typetexp.Error(_, _, _)
+
```

Comment:
I think the exception here is because typetexp directly `add_splice_lock` without checking? That makes type `'a` successfully parsed at stage `-1`, and only triggers exception later when you tries to print it.